soroban-env-host/src/host/metered_vector.rs

```diff
@@ -263,6 +263,21 @@ where
         }
     }
 
+    pub fn retain_mut<F>(&mut self, mut f: F, budget: &Budget) -> Result<Self, HostError>
+    where
+        F: FnMut(usize, &mut A) -> Result<bool, HostError>,
+    {
+        Self::charge_new(self.len(), budget)?;
+        let mut vec = Vec::with_capacity(self.len());
+        for (i, v) in self.vec.iter_mut().enumerate() {
+            if f(i, v)? {
+                vec.push(v.clone());
+            }
+        }
+        A::charge_for_clones(vec.as_slice(), budget)?;
+        Ok(Self { vec })
+    }
+
```

Comment:
@graydon Here is the `retain_mut` method we talked about. I couldn't make it work with iterator chaining or any other efficient iterator methods. But this approach also only does the allocation once. Let me know if you have a better way.